File: bresenham-circle.py
from pprint import pprint

class BresenhamCircleDrawing:
  def __init__(self, redius, center):
    self.redius = redius
    self.center = center
    self.circle_center_cordinate = [0, 0]
    self.circle = {
      "1st": [],
      "2nd": [],
      "3rd": [],
      "4th": [],
    }

  def run(self):
    currentX = self.circle_center_cordinate[0]
    currentY = self.redius

    tempCircle = {
      "1st": [],
      "2nd": [],
      "3rd": [],
      "4th": [],
    }

    currentP = 3 - 2 * self.redius

    first_octate = []
    second_octate = []

    while(currentX <= currentY):
      first_octate.append((currentX, currentY))
      second_octate.append((currentY, currentX))

      if currentP < 0:
        currentP += 4 * currentX + 6
        currentX += 1
      else:
        currentP += 4*(currentX - currentY) + 10
        currentX += 1
        currentY -= 1

    second_octate.pop()

    second_octate.reverse()

    """ first quadrant """
    tempCircle['1st'] = first_octate + second_octate  

    """ second quadrant """
    for x, y in tempCircle['1st']:
      tempCircle['2nd'].append((-x, y))

    """ third quadrant """
    for x, y in tempCircle['2nd']:
      tempCircle['3rd'].append((x, -y))
    
    """ fourth quadrant """
    for x, y in tempCircle['1st']:
      tempCircle['4th'].append((x, -y))

    

    """ 1st quadrant """
    for x, y in tempCircle['1st']:
      if y == 0: continue
      self.circle['1st'].append((x, y))

    """ 2nd quadrant """
    for x, y in tempCircle['2nd']:
      if x == 0: continue
      self.circle['2nd'].append((x, y))

    """ third quadrant """
    for x, y in tempCircle['3rd']:
      if y == 0: continue
      self.circle['3rd'].append((x, y))
    
    """ fourth quadrant """
    for x, y in tempCircle['4th']:
      if x == 0: continue
      self.circle['4th'].append((x, y))


    return self.circle
  

circle_radius = int(input("Enter circle radius === "))
circle_center = list(map(int, input("Enter circle radius === ").split()))

if(circle_radius < 0):
  raise ValueError("Circle radius must be positive")

pprint(BresenhamCircleDrawing(circle_radius, circle_center).run())

File: bresenham-circle.js
class BresenhamCircleDrawing {
  constructor(redius, centerCord = [0, 0]) {
    this.redius = redius;
    this.centerCordinate = centerCord;

    this.circle = {
      "1st": [],
      "2nd": [],
      "3rd": [],
      "4th": [],
    };
  }

  run() {
    let currentX = this.centerCordinate[0],
      currentY = this.redius;

    // let currentP = 1 - this.redius;
    let currentP = 3 - 2 * this.redius;

    let firstOctant = [];
    let secondOctant = [];

    while (currentX <= currentY) {
      firstOctant.push([currentX, currentY]);
      secondOctant.push([currentY, currentX]);

      if (currentP < 0) {
        // currentP += 2 * currentX + 3;
        currentP += 4 * currentX + 6;
        currentX++;
      } else {
        // currentP += 2 * (currentX - currentY) + 5;
        currentP += 4 * (currentX - currentY) + 10;
        currentX++;
        currentY--;
      }
    }

    secondOctant.pop();

    secondOctant = secondOctant.reverse();

    /* get first quadrant */
    this.circle["1st"] = [...firstOctant, ...secondOctant];

    /* get second quadrant */
    this.circle["2nd"] = this.circle["1st"].map(([x, y]) => [-1 * x, y]);

    /* get third quadrant */
    this.circle["3rd"] = this.circle["2nd"].map(([x, y]) => [x, -1 * y]);

    /* get fourth quadrant */
    this.circle["4th"] = this.circle["1st"].map(([x, y]) => [x, -1 * y]);


    this.circle["1st"] = this.circle["1st"].filter(([x, y]) => {
      return y !== 0;
    });

    this.circle["2nd"] = this.circle["2nd"].filter(([x, y]) => {
      return x !== 0;
    });

    this.circle["3rd"] = this.circle["3rd"].filter(([x, y]) => {
      return y !== 0;
    });

    this.circle["4th"] = this.circle["4th"].filter(([x, y]) => {
      return x !== 0;
    });

    return this.circle;
  }
}

console.log(new BresenhamCircleDrawing(7, [0, 0]).run());
